### ISSUE TITLE
Migrate to SFML2

### ISSUE THREAD

**Issue by Quix0r** [fabianschuiki/OpenSkyscraper#57]
I get this message with latest master:

```
[  1%] Building CXX object CMakeFiles/OpenSkyscraper.dir/source/PathFinder/GameMap.cpp.o
distcc[2656] ERROR: compile /home/bla/git/OpenSkyscraper/source/PathFinder/GameMap.cpp on 192.168.2.3 failed
distcc[2656] (dcc_build_somewhere) Warning: remote compilation of '/home/bla/git/OpenSkyscraper/source/PathFinder/GameMap.cpp' failed, retrying locally
distcc[2656] Warning: failed to distribute /home/bla/git/OpenSkyscraper/source/PathFinder/GameMap.cpp to 192.168.2.3, running locally instead
In file included from /home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../GUIManager.h:7:0,
                 from /home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../Application.h:10,
                 from /home/bla/git/OpenSkyscraper/source/PathFinder/../Item/Item.h:86,
                 from /home/bla/git/OpenSkyscraper/source/PathFinder/GameMap.cpp:2:
/home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../Rocket/SystemInterface.h:38:55: error: ‘sf::Key’ has not been declared
/home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../Rocket/SystemInterface.h:38:65: error: expected ‘,’ or ‘...’ before ‘Key’
In file included from /home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../Application.h:10:0,
                 from /home/bla/git/OpenSkyscraper/source/PathFinder/../Item/Item.h:86,
                 from /home/bla/git/OpenSkyscraper/source/PathFinder/GameMap.cpp:2:
/home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../GUIManager.h:23:55: error: ‘sf::Key’ has not been declared
/home/bla/git/OpenSkyscraper/source/PathFinder/../Item/../GUIManager.h:23:65: error: expected ‘,’ or ‘...’ before ‘key’
distcc[2656] ERROR: compile /home/bla/git/OpenSkyscraper/source/PathFinder/GameMap.cpp on localhost failed
CMakeFiles/OpenSkyscraper.dir/build.make:54: recipe for target 'CMakeFiles/OpenSkyscraper.dir/source/PathFinder/GameMap.cpp.o' failed
make[2]: *** [CMakeFiles/OpenSkyscraper.dir/source/PathFinder/GameMap.cpp.o] Error 1
CMakeFiles/Makefile2:60: recipe for target 'CMakeFiles/OpenSkyscraper.dir/all' failed
make[1]: *** [CMakeFiles/OpenSkyscraper.dir/all] Error 2
Makefile:76: recipe for target 'all' failed
make: *** [all] Error 2
```

Using this GCC version:
gcc version 4.9.2 (Debian 4.9.2-10) 

Maybe related to issue #51 ?

**Comment by matthewharvey** [comment]
From my understanding, this project was never really finished, and now depends on a library (SFML1.6) that isn't included with most distributions because it was superceded by the incompatible and more popular SFML2.
Since then there hasn't been much activity here.
Described in issue 51 is a way to build SFML1.6 off of the main system, and then reference it when compiling, so you might want to give that a try.
Also, porting is something that I've been trying on and off over the last little while, but it's rather arduous and I think it's pretty easy to screw up. A lot was changed between library versions. If you're interested in joining the effort (if only to review what I've already done) I'll put up an unfinished branch, but it still doesn't compile.

**Comment by Quix0r** [comment]
Okay, so currently no chance but to refacture the game to new API?

**Comment by piranna** [comment]
Seems so.

**Comment by fabianschuiki** [comment]
Yes I'm afraid a port to SFML2 would be the easiest way to get going again. Also, getting rid of Rocket and rolling a custom UI solution might be beneficial. @matthewharvey is working on an SFML2 port, which might already work. Unfortunately I don't have much time at my hands right now, but I hope to be able to setup packaging and deployment for Mac, Windows, and Linux at some point. The website is already up and running (openskyscraper.org).

**Comment by matthewharvey** [comment]
Just to be clear, my port is most certainly not complete. :(
And I'm pretty busy too right now, but I'm still monitoring these bug reports because the idea of this project is still something that interests me, so who knows.

**Comment by fabianschuiki** [comment]
@matthewharvey: I pulled your changes into an upstream branch and continued your work. There's quite a few things that have changed since sfml1.6, but it should be doable.

Maybe I can get things to build and pack properly on Mac, Windows, and Linux. That might lower the entry barrier for contributors to fix things.

**Comment by PCurd** [comment]
I'm not much of a C++ programmer (I work in C# mostly) but I loved SimTower and would very much be interested in helping - if it built etc. SFML2 is outside my experience level though.

**Comment by martinpfannemueller** [comment]
I also tried it multiple times to port it. In fact, only the SimTowerLoader is missing with the new branch. Since I already changed it, the project compiles but crashes on start with a segmentation fault. I am also not much of a C++ programmer. I will try to debug it. If there is interest in it, I can commit the changes, however I must say, that most changes are probably not the best solution. My primary goal was to simply let it compile.

**Comment by HoyteKashif** [comment]
Are you guys still working on this project? What is the current status?

**Comment by fabianschuiki** [comment]
I've merged and continued the efforts to port this to SFML2, currently happening in the `sfml2_port` branch. I've got it to properly compile and launch again on an Arch Linux setup. My current plan is to finalize this port and then setup [CPack](https://cmake.org/Wiki/CMake:Packaging_With_CPack) for Mac OS X, Windows, and Linux. This should make the project much more encouraging for contributors and would finally put the [website](http://www.openskyscraper.org/) to use again (it is quite stale at the moment).

**Comment by HoyteKashif** [comment]
I can program in C++ and i have a little experience with SFML so is there
any way that i can possibly help?

**Comment by fabianschuiki** [comment]
I think a nice start would be to check out the `sfml2_port` branch and see whether it works for you as well, and if not, to see what errors you can tackle that come up. I think I have overlooked a few things during the port, since the screen I currently get when launching the game lacks most of the textures and Rocket seems to be unable to properly texture things on screen. Getting this to work and look like the SFML1.6 version would be awesome.

Any help is highly appreciated :)

BTW: What platform are you working on?

**Comment by HoyteKashif** [comment]
I am working on Mac OSX 10.8 .

**Comment by fabianschuiki** [comment]
:+1: If you're having trouble getting your hands on libRocket (which seems to be pretty much a dead project), grab the [current `master` branch](https://github.com/libRocket/libRocket) of the project. They now have a script that installs the dependencies (`Dependencies/osx-depends.sh`) and a CMake build.

**Comment by HoyteKashif** [comment]
That is exactly what i did. What are some of the key things that still need
to be done?

**Comment by fabianschuiki** [comment]
Gameplay-wise we're about halfway through. About half of the items and means of transportation are implemented. The AI for visitors needs an overhaul. I would also suggest to move to C++11 and start to clean up the codebase by using some of the new idioms (smart pointers, etc.).

However, to continue on the gameplay stuff I guess we first need to fix the SFML2 port, such that we can actually see and debug changes to other parts of the code.

**Comment by fabianschuiki** [comment]
SFML2 has changed the Rect constructor from (left,top,right,bottom) to (left,top,width,height), which now of course causes a lot of trouble. Fix: Visit all calls to the constructor of Rect (i.e. IntRect and FloatRect) and subtract subtract the 1st from the 3rd and the 2nd from the 4th argument to map from the former to the latter representation. This should fix a portion of the rendering issues.

**Comment by HoyteKashif** [comment]
I am having trouble building the file system on MacOS platform. Is the TODO mentioned in the DataManager.cpp file current?

**Comment by fabianschuiki** [comment]
The TODO should not be an issue at the moment. `Path` is probably still broken for the cases mentioned therein, but it should not affect your build on Mac OS X. Things work for me on El Capitan. What troubles do you run into?

**Comment by HoyteKashif** [comment]
It is not creating the directories in Application Support. I get error
directories not found. I am also curious about the SImTower.exe file that I
am not sure how to create on a mac.

**Comment by fabianschuiki** [comment]
Actually, these are just informational messages. The console shows you a list of directories that are searched for the game's resources. However, the game does not create the Application Support directory, quite on purpose.

SimTower is available as abandonware, that is, you can download it (google for "SimTower abandonware"). That should give you a ZIP file with the contents of the original CD-ROM of the game. You now have two options: **1)** If you have the optional **libmspack** dependency installed, just copy the `SIMTOWER.EX_` to `~/Library/Application Support/OpenSkyscraper`. The game will automatically decompress the file. **2)** Otherwise you have to use a tool to extract the old Windows cabinet file, e.g. [cabextract](http://www.cabextract.org.uk/), and then place the decompressed `SIMTOWER.EXE` in the aforementioned directory.

**Comment by HoyteKashif** [comment]
Something tells me that there must be a better way.

**Comment by HoyteKashif** [comment]
Is there no way to change the method it uses to create the .exe and
decompression?

**Comment by fabianschuiki** [comment]
To avoid legal issues OpenSkyscraper has to depend on the player to supply the original game files. This way the player can be held liable for the means by which he acquires them. The resources of SimTower are embedded inside the executable, as was common practice back in the day. That means we need to open and parse the original `SIMTOWER.EXE` game, which is done in `SimTowerLoader` and `WindowsNELoader`. Now, most people don't have an installed version of SimTower, which is what you get if you run the `SETUP.EXE` on the game disc. Therefore OpenSkyscraper has the optional libmspack dependency that allows you to supply the compressed file `SIMTOWER.EX_` directly that can be found on the game disc, rather than having to run the setup program in some way.

I'm afraid this is the way we have to deal with the issue. libmspack and cabextract are both exceptionally well-programmed and should work like a charm.

**Comment by HoyteKashif** [comment]
Thank you for explaining.

**Comment by Quix0r** [comment]
OT: So why is there still a need for SimTower? How about getting rid of it all together? I mean games like FreeOrion do not depend on any old DOS/Win16 games (here: Master Of Orion) as they are independent and only "inspired" by the game.

**Comment by piranna** [comment]
To be able to play with the original data, textures, savegames... Except
for the last point, you could be able to generate your own game data files
with custom graphics if you want, but someone needs to do them.

**Comment by HoyteKashif** [comment]
I am all for that but I do not have the artistic skill to do so.

**Comment by fabianschuiki** [comment]
As @piranna said, the original game data is only needed to recreate SimTower's look and feel. I think the first step would be to get the game up and running with the original assets, since that amounts to the least effort (install [libmspack](http://www.cabextract.org.uk/libmspack/) and download download the [SimTower abandonware](http://www.abandonia.com/en/games/341/)). However, I do agree that later on, we should add another "enhanced" game mode that improves on the original SimTower by changing the mechanics and providing new, updated assets. Nevertheless, it's usually easier to add artwork once the core gameplay is setup (cf. [OpenTTD](http://www.openttd.org)).

**Comment by Quix0r** [comment]
Like the last poster's idea. :+1:

**Comment by AzemOcram** [comment]
I will try to update my OcramTower graphics pack in December. I stopped because I thought this project died and I transferred into a full-time ChemE major. I started experimenting with Inkscape in August but only vectorized 7 of my rooms by hand. The PNG renders of my SVG art looks better than the originals (made in MS Paint and Gimp cerca 2010)

**Comment by AzemOcram** [comment]
FYI: here is the project with my attempt to make an HTL5+ECMAScript game to show off my art: https://github.com/AzemOcram/OcramTower
http://azemocram.github.io/OcramTower/

**Comment by piranna** [comment]
I love the idea of the html version :-D But why don't use directly the SVG images? Instead of using a Canvas you can use a (big) <svg> tag and include there the SVG image of the rooms, thus would allow you to have animated rooms instead of static ones and also you would click events to interact with them in Javascript.

**Comment by AzemOcram** [comment]
Well, the images in OcramTower were all pixel (pushed) art. I didn't start redoing them as SVG until this past summer. I started making the SVG because the Canvas tag was confusing.

**Comment by AzemOcram** [comment]
FYI, I just committed the SVG images to the master branch of my project a couple minutes before this post.

Here: https://github.com/AzemOcram/OcramTower/tree/master/Rooms/Vector

**Comment by piranna** [comment]
Nice images :-)

**Comment by Quix0r** [comment]
@AzemOcram Oh, a bit OT here but please make Index.html lower-case, else Apache under Linux cannot load it.

**Comment by kirtangajjar** [comment]
Any updates?

**Comment by fabianschuiki** [comment]
It's pretty tedious work. A lot of things have changed for SFML2, and the code base itself is not very clean and easy to use. I'm tempted to scrap most of it and do a clean re-write, rendering directly in OpenGL and also dumping libRocket for the GUI (which is fairly cumbersome and adds more hassle than it brings comfort).

**Comment by PCurd** [comment]
Sorry if I missed an earlier discussion on this but is there any reason it
needs to be written using OpenGL etc - could it be ported to an "off the
shelf" engine like Unity?

**Comment by vladimir1986** [comment]
A good reason not to use Unity is to keep this game multiplatform. Unity=goodbye to Linux and Unix.

I really am up to port it to OpenGL. Using so many obscure libraries is cumbersome to install and run in to the risk of eventualliy having to use an unsupported, ancient library, or needing to change the complete code of the nearly finished game later... I am a strong believer of multiplatform, and as less as possible libraries.

**Comment by PCurd** [comment]
Unity is hugely multiplatform - https://unity3d.com/unity/multiplatform

I play numerous Unity games on Mac (RimWorld being the first that comes to
mind).

**Comment by matthewharvey** [comment]
Using Unity would satisfy the multi platform thing. It would also put this project at the whims of Unity the company, from them fixing bugs that ail us to them updating their licencing terms to tell us that we can no longer distribute unlimited copies (assuming that their licencing terms would currently be useful for a project like this).

**Comment by AzemOcram** [comment]
OpenGL would be best and it would let OpenSkyscraper do more with OpenGL  than it ever possibly could with SFML.

**Comment by piranna** [comment]
Being a 2D game, better stick with a 2D library/engine/technology like SVG
or Cairo.

**Comment by AzemOcram** [comment]
OpenGL works with 2D just as good as Unity does and both can do 3D. I have toyed around with the student version of Unity (trying to make a cross between Monopoly and SimCity, ended up making a physical board game). OpenGL works great with sprites and 2D images and doesn't require depth (or maybe Z has to be set to 0). Highrise Developer (the SimTower clone that previously went the farthest) used OpenGL in order to use images of different resolutions (24, 48, 72, 96, and even one 192 pixels per storey image) and smooth zooming. However, most skyscrapers are more than one room deep and OpenGL can easily let the developers (or modders, the line blurs in an open source game) add depth later on. Rooms can still be rendered as 2D pictures but I have played OpenGL games with multiple views/angles. My idea for an offline, standalone SimTower clone is to first add all the functionality of SimTower, then make better art (which I already did) and more functionality (the good parts of Yoot Tower plus some ideas we brainstormed since 2008, some of which will go into SomaSim's Project Highrise), add depth and a top-down view, add isometric view, go full 3D with even more features.

Too bad my programming skills are limited so I can barely make programs that let me sandbox (no simulation, money, or restraints) build skyscrapers with my art.

**Comment by fabianschuiki** [comment]
Just to clarify such that we don't run into any misunderstandings:

The question is not whether to use **OpenGL** or not. OpenGL is one of the two common APIs to talk to your graphics card for rendering, the other being DirectX. Of the two, only OpenGL is multi-platform, and since a hard requirement for the game is to run on Linux, Mac, and Windows, this is the only option.

**Unity** on the other hand is a 3D engine that sits ontop of OpenGL and handles the communication with the graphics hardware, such that game developers can work on a higher level of abstraction. In doing so, Unity is good at most at what it does, but not brilliant, since its generality requires the engine to make a few unfavorable tradeoffs that would not be necessary in a dedicate engine. Incidentally, this is why studios developing anything but ego shooters tend to roll their own engine, e.g. for open world games such as GTA, Skyrim, or Fallout.

**SFML** is a library that acts as an abstraction layer between the underlying platform and the game, to allow for quick prototyping. It handles one of the more annoying parts of game development: Opening an OpenGL context for drawing (a window, or fullscreen), which is highly platform-dependent. Additionally, SFML provides a few classes that issue OpenGL rendering commands (e.g. sprites) to get a few initial things on screen.

**SVG** and **Cairo** are technologies developed for the web and desktop UIs respectively. SVG is simply a file format for vector graphics that can easily be used to draw content in web browsers. However, vector graphics are inherently slow to draw. Cairo is an abstract API for drawing images and shapes at high quality (i.e. nice anti-aliasing) and is generally modelled after Apple's Quartz API. This makes it very suitable for drawing static UIs for desktop applications (e.g. the buttons in Photoshop) that can be buffered and need not be refreshed for a long time, but much too slow for a game since its main backend was designed to draw on the CPU rather than on the GPU.

The bottom line is that Unity would be an option, but is probably too big for a game as simple as OpenSkyscraper, and we are already using OpenGL through SFML's abstraction. SVG and Cairo are no options since they were designed for completely different applications and are too slow. We *could* use SVG to store sprites on disk and render them to the desired resolution once at launch.

The problem with OpenGL is that API calls, especially calls that issue draw commands to the GPU, are very, very slow since they have to make their way from the game to the OpenGL library, into kernel mode, into the graphics card driver, into the PCI driver, across the PCI bus (which is fairly slow) into the GPU. For each sprite, SFML simply issues an OpenGL draw call, which incurs an incredible overhead. If we were to write a proper engine that groups sprites up and sends them in batches (many sprites per draw call), we could easily draw 1000 sprites during the time it takes to draw a single SFML sprite. Incidentally, this is one of the tasks a game engine does for you: grouping up drawables and ordering/preparing them for efficient dispatch.

So dispatching our own OpenGL calls and writing our own sprites would allow **a)** for cleaner and simpler code, **b)** much faster execution times, and **c)** less reliance on SFML in case we want to exchange it for the much more streamlined GLFW3 later on. The same holds for the GUI: Getting rid of libRocket and writing our own GUI library would remove the dependency on a very tedious library and probably be faster.

**tl;dr**: Let's rewrite the graphics code and GUI in OpenGL 4 for a modern, streamlined, and fast game.

**Comment by piranna** [comment]
> SVG and Cairo are technologies developed for the web and desktop UIs respectively. SVG is simply a file format for vector graphics that can easily be used to draw content in web browsers. However, vector graphics are inherently slow to draw. Cairo is an abstract API for drawing images and shapes at high quality (i.e. nice anti-aliasing) and is generally modelled after Apple's Quartz API. This makes it very suitable for drawing static UIs for desktop applications (e.g. the buttons in Photoshop) that can be buffered and need not be refreshed for a long time, but much too slow for a game since its main backend was designed to draw on the CPU rather than on the GPU.
> 
> The bottom line is that Unity would be an option, but is probably too big for a game as simple as OpenSkyscraper, and we are already using OpenGL through SFML's abstraction. SVG and Cairo are no options since they were designed for completely different applications and are too slow. We could use SVG to store sprites on disk and render them to the desired resolution once at launch.

I must to HEAVILY DISAGREE here. I have work developing a SVG engine 7 years ago capable of >450fps @ 1080p, I did an OpenGL 3D engine capable of moving 3.500.000 polygons with transparency on a laptop ATI Radeon 3200 mobile graphic card (from 10 years ago) and I've also ported node-canvas (based on Cairo) to the Linux framebuffer, so I know what I'm saying: SVG is not slow at all, specially by using libraries like Cairo that support hardware acceleration both by mapping OpenGL calls or by using OpenVG is supported by the graphic card. Vectorial graphics are not slow by themselves, as you told Cairo is inspired by Apple Quartz, that at the same time is based on PDF-X/PostScript (one of the first vectorial graphic systems). Do you think the OSX interface is slow? Or the fact that SVG is used as replacement for Flash animations? I don't think so... I think SVG and Cairo are totally valid technologies for this project.


> tl;dr: Let's rewrite the graphics code and GUI in OpenGL 4 for a modern, streamlined, and fast game.

Oh, cool, let's use a high-edge technology only available about 3-5 years ago on high-end computers (mine don't support anything beyond OpenGL 2.1...) to mimic a game from 20 years ago just to make it "fast"...

**Comment by AzemOcram** [comment]
Thank you so much for the lesson. As an amateur/hobbyist artist and web designer, I didn't know much about programming. I will gladly make all the art for your game once you decided on the style (resolution/detail, decade, country) and format (raster vs vector).

**Comment by HoyteKashif** [comment]
Fabianschuiki, I was wondering what tools you used if any to reverse engineer the zip file? I have gone through the code in the simloader.cpp/.h but want to know what led you to know the location of each resource and their size? Thanks

**Comment by rodneyrod** [comment]
Well, as a non-coder, is there anything that we can do to help like test new code?

**Comment by fabianschuiki** [comment]
@HoyteKashif I started with the game executable that you get after you run the installer. It is a Windows NE executable, for which you can find some, although not much, information on the web that allows you to figure out where the resources are: There's an MSDOS header that points to an NE header, that points to a resource table, that in turn points to arrays of resources of a specific type, which in turn contains descriptors for individual bitmaps and sounds, which in turn contain pointers to their actual data. Adding support for the executable before installation, or the ZIP archive, is fairly easy: There's [libmspack](http://www.cabextract.org.uk/libmspack/) to extract `SIMTOWER.EX_`, and there's [libzip](https://nih.at/libzip/) to extract ZIP files.

**Comment by fabianschuiki** [comment]
@rodneyrod At the moment the possibilities are limited, since the SFML2 rewrite takes some time and there is little to test: It either looks right, or it doesn't. However once this is done, help with testing code is highly appreciated. And of course there's always the possibility to come up with a new art style in case we want to offer an option to play the game without the original SimTower, but with new and enhanced visuals.

**Comment by hellcp** [comment]
@fabianschuiki count me in as interested in making new assets, [I packaged the game for openSUSE](https://build.opensuse.org/package/show/home:hellcp:branches:games/openskyscraper) considering that it almost works right with SFML2 (there are a few crashes here and there), however I cannot package simtower.exe due to copyright, obviously. It's easy enough to add the file post install, but it's still not desired, it would be easier to have everything open source :D

I can do graphics for that no problem, I would just need some pointers as to what exactly is needed :P

**Comment by Quix0r** [comment]
Back from silence. :-) The `TODO.md` file still states `Maybe we should move everything to SFML version 2.0.` but that seems to be done. :+1: And the compilation error is gone (compiles and links completely, **you may have to cleanup your "build" directory"**).

**Comment by fabianschuiki** [comment]
@hellcp Thanks a lot for packaging the project. Having separate, tailored graphics would be very nice and appreciated. I don't have a lot of time to contribute to the project myself unfortunately. But I'm very happy about contributions 😃 !

**Comment by fabianschuiki** [comment]
@Quix0r I think you're right. Time to close the issue and adjust the TODO.md file 👍

**Comment by Quix0r** [comment]
So thank you all for porting and fixing it. :-)

**Comment by hellcp** [comment]
@fabianschuiki would be good to update README too, so it reflects that `Restart` branch is no longer a thing
